### PR TITLE
feat(ui): persist article draft to prevent data loss on navigation

### PR DESCRIPTION
**Pull request by Rohit0301**
### Describe your changes:

I added a localStorage-backed Zustand store (`useArticleDraftStore`) that captures article content and title every 300 ms while the user types. If the user navigates away or refreshes before the 3–10 s server debounce fires, the draft is preserved. On return to the same article, the skeleton blocks editing until the draft is auto-synced to the server and then cleared.

Two related bugs are also fixed: a stale debounce from Article A could overwrite Article B's description/version in state after navigation (`setKnowledgePage` is now guarded by article ID), and the "Unsaved" badge was bleeding into the next article because `contentChangeState` was never reset on `fqn` change.

#
### Type of change:
- [ ] Bug fix
- [x] Improvement
- [ ] New feature
- [ ] Breaking change
- [ ] Documentation

#
### High-level design:

**New store** (`hooks/useArticleDraftStore.ts`): Zustand `persist` store (key `om-article-drafts`, localStorage) keyed by article ID — supports multiple drafts simultaneously.

**Draft lifecycle in `KnowledgePageDetailComponent`:**
- On keystroke → `setDraft` via 300 ms debounce (avoids per-keystroke localStorage writes)
- On successful server save → `removeDraft` (draft no longer needed)
- On page load with pending draft → apply draft to editor, inline-await `patchKnowledgePage`, keep skeleton visible throughout so user can't make concurrent edits; clear draft on success
- On fqn change → reset `contentChangeState` and `pendingSaveCountRef` so Article B starts clean
- Stale debounce guard → `isCurrentArticle` flag (captured before `await`) prevents Article A's server response from overwriting Article B's state or triggering its save badge

**Deletion cleanup**: `removeDraft` called after successful delete in `ArticleDetailHeader`, `KnowledgePagesHierarchy`, and `KnowledgeCard`.

#
#### Use cases covered
- User types in an article, refreshes before auto-save fires → draft is restored and synced on reload
- User edits two articles and navigates between them → each article's draft is tracked independently
- User deletes an article that has a pending draft → draft is cleared from localStorage
- User edits Article A, switches to Article B within the debounce window → Article B shows its own content and "Saved" badge (not Article A's)

#### Unit tests
- [ ] I added unit tests for the new/changed logic.

#### Backend integration tests
- [x] Not applicable (no backend API changes).

#### Ingestion integration tests
- [x] Not applicable (no ingestion changes).

#### Playwright (UI) tests
- [ ] I added Playwright E2E tests under `openmetadata-ui/.../ui/playwright/` for UI changes.

#### Manual testing performed
1. Opened an article, typed content, refreshed within 3 seconds → skeleton showed, content was saved and displayed correctly
2. Edited two articles rapidly, navigated between them → no cross-article content bleed, badges correct
3. Deleted an article with unsaved changes → localStorage key cleared

#
### UI screen recording / screenshots:

Not applicable.

#
### Checklist:
- [x] I have read the [**CONTRIBUTING**](https://docs.open-metadata.org/developers/contribute) document.
- [ ] My PR title is `Fixes <issue-number>: <short explanation>`
- [ ] My PR is linked to a GitHub issue via `Fixes #<issue-number>` above.
- [ ] I have commented on my code, particularly in hard-to-understand areas.
- [x] For JSON Schema changes: N/A
- [ ] For UI changes: I attached a screen recording and/or screenshots above.
- [ ] I have added tests (unit / integration / Playwright as applicable) and listed them above.

<details open><summary><h3>Greptile Summary</h3></summary>

This PR adds localStorage-backed draft persistence for Knowledge Center articles to prevent data loss when the user navigates away before the server-side auto-save debounce fires. It also fixes two related bugs: a stale debounce from Article A could overwrite Article B's state after navigation, and the "Unsaved" badge was not reset on `fqn` change.

- **New `useArticleDraftStore`**: Zustand `persist` store (key `om-article-drafts`) keyed by article ID; a 300 ms debounce writes description/displayName/version to localStorage on every keystroke, and `removeDraft` is called on successful save or article deletion from all three delete sites.
- **Per-article pending-save tracking**: `pendingSaveCountByArticleRef` replaces the single `pendingSaveCountRef`; `beginTrackedSave`/`endTrackedSave` now receive an explicit `savedArticleId` captured in the closure, so a save that resolves after navigation cannot update the badge or remove the draft of the currently-displayed article.
- **Draft sync on page load**: `fetchKnowledgePage` checks for a pending draft, applies it to the editor, fires a PATCH while the skeleton is still visible, then clears the draft on success; the `fqn` effect resets `contentChangeState` to SAVED so Article B always starts clean.
</details>

<details open><summary><h3>Confidence Score: 5/5</h3></summary>

Safe to merge — the previous-thread issues around stale badge updates and wrong-article draft removal have been addressed; the remaining edge case is a UX gap, not a data-corruption path.

The cross-article badge bleed and wrong-`removeDraft` bugs flagged in earlier reviews are resolved: `savedArticleId` is captured in each save closure, `knowledgePageIdRef.current` is checked at resolution time, and per-article pending-save tracking prevents stale saves from affecting the currently displayed article. The only outstanding item is that a draft is silently discarded when a concurrent edit bumps the server version — a niche race condition that produces no incorrect server state, only a missed UX notification.

The draft-sync path in `KnowledgePageDetailComponent.tsx` (lines 193–228) deserves a second look for the silent discard case; all other files are straightforward.
</details>

<details open><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| openmetadata-ui/src/main/resources/ui/src/hooks/useArticleDraftStore.ts | New Zustand persist store keyed by article ID — `setDraft`, `removeDraft`, and `getDraft` are correctly implemented; `removeDraft` is a no-op when the key is absent, preventing spurious localStorage writes. |
| openmetadata-ui/src/main/resources/ui/src/components/KnowledgeCenter/KnowledgePageDetailComponent/KnowledgePageDetailComponent.tsx | Core draft lifecycle logic added: per-article pending-save tracking, 300 ms draft debounce, fqn-change badge reset, and draft-sync on load. Previous-thread issues (stale `setContentChangeState` for wrong article, `removeDraft` removing wrong draft) are now addressed by the `savedArticleId` capture + `knowledgePageIdRef` guard in `endTrackedSave`. One remaining edge case: unsaved draft silently discarded when `serverChangedSinceDraft` is true. |
| openmetadata-ui/src/main/resources/ui/playwright/e2e/Features/ContextCenterArticles.spec.ts | Adds six E2E tests covering draft persistence, cross-article isolation, reload sync, no-spurious-PATCH on clean reload, and delete-clears-draft. Tests look correct; articles are created via API in beforeAll and cleaned up. |
| openmetadata-ui/src/main/resources/ui/playwright/utils/ContextCenterUtil.ts | Adds `readDraftStore` helper and improves `scrollHierarchyToNode` stability (3-consecutive-stale guard, 100 iterations, 5 s timeout). The localStorage key is hardcoded separately from the store definition — minor coupling risk. |
| openmetadata-ui/src/main/resources/ui/src/components/ContextCenter/ArticleDetailHeader/ArticleDetailHeader.component.tsx | Removes the now-unused explicit Save button and `showSaveButton` variable; adds `removeDraft` call on successful delete. Change is clean and well-scoped. |
| openmetadata-ui/src/main/resources/ui/src/components/KnowledgeCenter/KnowledgeCard/KnowledgeCard.tsx | Adds `removeDraft(knowledgePage.id)` on successful card-level delete — mirrors the same cleanup added to `ArticleDetailHeader` and `KnowledgePagesHierarchy`. |
| openmetadata-ui/src/main/resources/ui/src/components/KnowledgeCenter/KnowledgePagesHierarchy/KnowledgePagesHierarchy.tsx | Adds `removeDraft(deletePage.id)` after hierarchy-level delete. Consistent with the same cleanup in other delete paths. |
| openmetadata-ui/src/main/resources/ui/src/pages/ContextCenterPage/ContextCenterArticlesPage/ContextCenterArticlesPage.tsx | Removes the `onSave` prop pass-through (the explicit Save button is gone from `ArticleDetailHeader`). Minimal, correct change. |

</details>

<details open><summary><h3>Sequence Diagram</h3></summary>

<a href="#gh-light-mode-only">

```mermaid
%%{init: {'theme': 'neutral'}}%%
sequenceDiagram
    participant U as User
    participant KC as KnowledgePageDetailComponent
    participant DS as useArticleDraftStore (localStorage)
    participant API as Server API

    U->>KC: Type in Article A editor
    KC->>DS: setDraft(A.id) — 300 ms debounce
    KC->>API: patchKnowledgePage(A.id) — 3-10 s debounce
    U->>KC: Navigate to Article B before server save fires
    KC->>KC: fqn effect → setContentChangeState(SAVED)
    KC->>API: getKnowledgePageByFqn(B.fqn)
    Note over KC: knowledgePageIdRef.current → B.id after render
    API-->>KC: Article A PATCH resolves (stale)
    KC->>KC: "savedArticleId(A) != knowledgePageIdRef(B) → no badge update"
    KC->>DS: removeDraft(A.id)
    U->>KC: Reload Article A
    KC->>API: getKnowledgePageByFqn(A.fqn)
    API-->>KC: response
    KC->>DS: getDraft(A.id)
    DS-->>KC: draft with description and version
    KC->>KC: "hasChanges=true serverChangedSinceDraft=false"
    KC->>KC: setKnowledgePage(pageWithDraft) + skeleton shown
    KC->>API: patchKnowledgePage(A.id, draftPatch)
    API-->>KC: saved response
    KC->>DS: removeDraft(A.id)
    KC->>KC: setContentChangeState(SAVED) + skeleton hidden
```

</a>
<a href="#gh-dark-mode-only">

```mermaid
%%{init: {'theme': 'base', 'themeVariables': {"darkMode": true, "background": "#0d1117", "primaryColor": "#21262d", "primaryTextColor": "#e6edf3", "primaryBorderColor": "#8b949e", "lineColor": "#8b949e", "textColor": "#e6edf3", "edgeLabelBackground": "#161b22", "actorBkg": "#21262d", "actorBorder": "#8b949e", "actorTextColor": "#e6edf3", "actorLineColor": "#8b949e", "signalColor": "#8b949e", "signalTextColor": "#e6edf3", "noteBkgColor": "#373320", "noteBorderColor": "#d4a72c", "noteTextColor": "#f0e6c0", "labelBoxBkgColor": "#21262d", "labelBoxBorderColor": "#8b949e", "labelTextColor": "#e6edf3", "loopTextColor": "#e6edf3", "activationBkgColor": "#30363d", "activationBorderColor": "#8b949e"}}}%%
sequenceDiagram
    participant U as User
    participant KC as KnowledgePageDetailComponent
    participant DS as useArticleDraftStore (localStorage)
    participant API as Server API

    U->>KC: Type in Article A editor
    KC->>DS: setDraft(A.id) — 300 ms debounce
    KC->>API: patchKnowledgePage(A.id) — 3-10 s debounce
    U->>KC: Navigate to Article B before server save fires
    KC->>KC: fqn effect → setContentChangeState(SAVED)
    KC->>API: getKnowledgePageByFqn(B.fqn)
    Note over KC: knowledgePageIdRef.current → B.id after render
    API-->>KC: Article A PATCH resolves (stale)
    KC->>KC: "savedArticleId(A) != knowledgePageIdRef(B) → no badge update"
    KC->>DS: removeDraft(A.id)
    U->>KC: Reload Article A
    KC->>API: getKnowledgePageByFqn(A.fqn)
    API-->>KC: response
    KC->>DS: getDraft(A.id)
    DS-->>KC: draft with description and version
    KC->>KC: "hasChanges=true serverChangedSinceDraft=false"
    KC->>KC: setKnowledgePage(pageWithDraft) + skeleton shown
    KC->>API: patchKnowledgePage(A.id, draftPatch)
    API-->>KC: saved response
    KC->>DS: removeDraft(A.id)
    KC->>KC: setContentChangeState(SAVED) + skeleton hidden
```

</a>
</details>

<details><summary><h3>Comments Outside Diff (1)</h3></summary>

1. `openmetadata-ui/src/main/resources/ui/src/components/KnowledgeCenter/KnowledgePageDetailComponent/KnowledgePageDetailComponent.tsx`, line 370-378 ([link](https://github.com/open-metadata/openmetadata/blob/67a0776e907357e684250216aa91c36c2222fcb7/openmetadata-ui/src/main/resources/ui/src/components/KnowledgeCenter/KnowledgePageDetailComponent/KnowledgePageDetailComponent.tsx#L370-L378)) 

   <a href="#">P1</a> **`endTrackedSave` removes the wrong article's draft after navigation**

   `knowledgePageIdRef.current` is read at call time, not at save-start time. If the user navigates from Article A to Article B while Article A's save is in-flight, the `fqn` effect resets `pendingSaveCountRef.current` to 0 and the `knowledgePage.id` effect updates `knowledgePageIdRef.current` to Article B's ID. When Article A's save later resolves and calls `endTrackedSave`, the counter is already 0, so `Math.max(0, 0-1)` stays at 0, the `pendingSaveCountRef.current === 0` branch is entered, `setContentChangeState(SAVED)` silently clears Article B's unsaved badge, and `removeDraft(knowledgePageIdRef.current)` removes Article B's draft instead of Article A's. The `isCurrentArticle` flag captured before the await is always `true` for the article being saved (both values are Article A's ID at capture time), so it does not prevent this path from executing. The draft ID used in `endTrackedSave` must be captured at save-start time and closed over, not read from the ref at finalization time.
</details>

<sub>Reviews (7): Last reviewed commit: ["Merge branch &#39;main&#39; into article-unsaved..."](https://github.com/open-metadata/openmetadata/commit/4e68e126c5dfb785a3df0dee82da8bb68e6d3f73) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=42144302)</sub>

### CODE REVIEW

File: openmetadata-ui/src/main/resources/ui/src/components/KnowledgeCenter/KnowledgePageDetailComponent/KnowledgePageDetailComponent.tsx [hunk 1]
```diff
@@ -175,7 +178,58 @@ const KnowledgePageDetailComponent: FC<KnowledgePageDetailComponentProps> = ({
...
+          }
+        } catch (syncError) {
+          showErrorToast(syncError as AxiosError);
+          if (response.id === knowledgePageIdRef.current) {
+            setContentChangeState(ContentChangeState.UN_SAVED);
+          }
+        }
+      } else {
+        setKnowledgePage(response);
+        if (draft) {
```
Comment (chirag-madlani) [inline openmetadata-ui/src/main/resources/ui/src/components/KnowledgeCenter/KnowledgePageDetailComponent/KnowledgePageDetailComponent.tsx]: Do we need this check?

**Review by chirag-madlani — APPROVED** [review]
